storage/metrics.go

```diff
@@ -0,0 +1,187 @@
+	MaxChannelsInReport     = 50
+	EntriesInReportByDay    = 14 // days report returns previous 2 weeks etc
+	EntriesInReportByHour   = 24
+	EntriesInReportByMinute = 30
+)
+
+type UserMetrics struct {
+	es        *elasticsearch.Client
+	indexName string
+}
+
+func NewUserMetrics(es *elasticsearch.Client, indexName string) *UserMetrics {
+	return &UserMetrics{es, indexName}
+}
+
+type UserActivityAggs struct {
+	Channels struct {
+		Terms struct {
+			Field string `json:"field"`
+			Size  int    `json:"size"`
+		} `json:"terms"`
+		Aggs *DateHistogramAggs `json:"aggs,omitempty"`
+	} `json:"channels"`
+}
+
+type DateHistogramAggs struct {
+	BySpecifiedRange struct {
+		DateHistogram struct {
+			Field            string `json:"field"`
+			CalendarInterval string `json:"calendar_interval"`
+			MinDocCount      int    `json:"min_doc_count"`
+			Format           string `json:"format"`
+			ExtendedBounds   struct {
+				Min int64 `json:"min"`
+				Max int64 `json:"max"`
+			} `json:"extended_bounds"`
+			HardBounds struct {
+				Min int64 `json:"min"`
+			} `json:"hard_bounds"`
+		} `json:"date_histogram"`
+	} `json:"by_specified_range"`
+}
+
+type dateBucket struct {
+	KeyString string `json:"key_as_string"`
+	DocCount  int    `json:"doc_count"`
+}
+
+type channelBucket struct {
+	Key              string `json:"key"`
+	DocCount         int    `json:"doc_count"`
+	BySpecifiedRange *struct {
+		Buckets []dateBucket `json:"buckets"`
+	} `json:"by_specified_range,omitempty"`
+}
+type channels struct {
+	Buckets []channelBucket `json:"buckets"`
+}
+type aggs struct {
+	Channels channels `json:"channels"`
+}
+type queryResult struct {
+	Aggregations aggs `json:"aggregations"`
+}
+
+func (s *UserMetrics) runMessageAggregation(query io.Reader) (*queryResult, error) {
+	res, err := s.es.Search(
+		s.es.Search.WithContext(context.Background()),
+		s.es.Search.WithIndex(s.indexName),
+		s.es.Search.WithBody(query),
+		s.es.Search.WithTrackTotalHits(true),
+		s.es.Search.WithPretty(),
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+	queryResult, err := unmarshalMessageAggregationQueryResult(res)
+	if err != nil {
+		return nil, err
+	}
+	return queryResult, nil
+}
+
+// This report returns # of messages by user in channel.
+// Additional grouping by day/hour/minute is possible.
+// If groupBy option is provided, the amount of returned data is limited according to constants in this module.
+func (s *UserMetrics) GetUserMessagesCounts(opts *api.UserActivityReportOptions) (*api.UserActivityReport, error) {
+	query := buildUserActivityQuery(opts)
+	query_result, err := s.runMessageAggregation(query)
+	if err != nil {
+		return nil, api.ErrInternal
+	}
+
+	results := make(api.UserActivityReport)
+	for _, hit := range query_result.Aggregations.Channels.Buckets {
+		activityItem := &api.ChannelActivity{TotalMessages: hit.DocCount}
+		if hit.BySpecifiedRange != nil {
+			activityItem.MessageCountDistribution = []api.DistributionEntry{}
```

Comment:
```suggestion
			activityItem.MessageCountDistribution = make([]api.DistributionEntry, 0, len(hit.BySpecifiedRange.Buckets))
```
Slice capacity can be allocated ahead of time

Reply:
done in [apply review suggestions](https://github.com/kacperf531/sockchat/pull/11/commits/4054b76b6854361e07ae3021db901a9e9bd32cb8)